src/subscription_manager/repofile.py

```diff
@@ -469,6 +473,11 @@ def fix_content(self, content):
             apt_cont['Suites'] = 'default'
             apt_cont['Components'] = 'all'
             apt_cont['Trusted'] = 'yes'
+
+            if 'arches' in apt_cont and apt_cont['arches'] != '0':
+                arches_str = " ".join(apt_cont['arches'])
```

Comment:
This doesn't work, when I use Python3. This line produces following traceback:

```
2020-11-24 13:19:16,437 [ERROR] subscription_manager.py:44196:MainThread @managercli.py:220 - sequence item 0: expected str instance, bytes found
Traceback (most recent call last):
  File "/home/jhnidek/github/candlepin/subscription-manager/src/subscription_manager/managercli.py", line 547, in main
    return_code = self._do_command()
  File "/home/jhnidek/github/candlepin/subscription-manager/src/subscription_manager/managercli.py", line 2694, in _do_command
    rc = self._set_repo_status(repos, rl, self.options.repo_actions)
  File "/home/jhnidek/github/candlepin/subscription-manager/src/subscription_manager/managercli.py", line 2779, in _set_repo_status
    repo_action_invoker.update()
  File "/home/jhnidek/github/candlepin/subscription-manager/src/subscription_manager/certlib.py", line 32, in update
    self.report = self.locker.run(self._do_update)
  File "/home/jhnidek/github/candlepin/subscription-manager/src/subscription_manager/certlib.py", line 18, in run
    return action()
  File "/home/jhnidek/github/candlepin/subscription-manager/src/subscription_manager/repolib.py", line 195, in _do_update
    res = action.perform()
  File "/home/jhnidek/github/candlepin/subscription-manager/src/subscription_manager/repolib.py", line 431, in perform
    fixed_cont = repo_file.fix_content(cont)
  File "/home/jhnidek/github/candlepin/subscription-manager/src/subscription_manager/repofile.py", line 478, in fix_content
    arches_str = " ".join(apt_cont['arches'])
TypeError: sequence item 0: expected str instance, bytes found
```

I would fix it with something like this:

```python
arches = apt_cont['arches']
if len(arches) > 0 and type(arches[0]) == bytes:
    arches = [arch.decode('utf-8') for arch in arches]
arches_str = " ".join(arches)
```

Reply:
The problem is the different representation of strings in Python2 and Python3.

Starting at [line 157](https://github.com/candlepin/subscription-manager/blob/7ebf96eaf1edd53efa5f2ea0efdb082ef2d66e95/src/subscription_manager/repofile.py#L157) where the `arches` attribute is filled with data from the cert content. The data arrives as *unicode* string which is the default string type for Python3, but not for Python2. Accordingly, in Python2 the `x.encode('utf-8')` part makes the string "normal". As for Python3, this line changes the string from the "normal" type *unicode* to *bytes*.
So, any suggestions on how to differ between Python2 and 3 at this point? I think we must leave the data as *unicode* when running Python3 and change the type from *unicode* to *bytes* when running Python2.

Reply:
@jirihnidek I just added a line to determine whether the given string is in the *normal* format of the running Python version. Therefore, I use `isinstance(my_str, "".__class__)`.